File: routing-v4/src/orchestrator.py
"""
orchestrator.py  —  routing-v4
================================
Ties together: splitting → routing (with failover) → parallel execution
→ telemetry recording → feedback persistence → WebSocket broadcast.

Every completed sub-task emits a routing_event to the async event queue
so the front-end receives a real-time update for each individual LLM call.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Optional

from src.splitter import split_and_classify
from src.router import dispatch, _utility, STATIC_TASK_THRESHOLDS, BASE_QUALITY_THRESHOLD
from src.llm_client import call_model, TIER_TO_MODEL
from src.feedback import FeedbackStore
from src.failover_manager import failover_manager, FailoverEvent
from src.telemetry import telemetry_engine, TelemetryRecord
from src.task_classifier import TaskProfile
from src.cost_registry import ModelSpec


@dataclass
class SubTaskResult:
    profile:        TaskProfile
    model:          ModelSpec
    response:       str
    latency_ms:     int
    out_tokens:     int
    # Routing metadata
    utility_score:  float
    quality_score:  float
    threshold:      float
    ollama_model:   str
    cost_units:     float
    # Failover metadata
    failover_used:  bool              = False
    failover_from:  str               = ""
    failover_reason: str              = ""
    # Dynamic params that were active when this call was routed
    dyn_threshold:  float             = 0.0
    dyn_cost_weight: float            = 0.0


@dataclass
class OrchestratorResult:
    prompt:      str
    sub_results: list[SubTaskResult]
    merged:      str
    total_ms:    int

    def routing_table(self) -> list[dict]:
        return [
            {
                "task_type":      r.profile.task_type.value,
                "complexity":     r.profile.complexity,
                "tier":           r.profile.tier.name,
                "model":          r.model.label,
                "provider":       r.model.provider,
                "ollama_model":   r.ollama_model,
                "utility":        round(r.utility_score, 4),
                "quality":        round(r.quality_score, 4),
                "threshold":      r.threshold,
                "latency_ms":     r.latency_ms,
                "out_tokens":     r.out_tokens,
                "cost_units":     r.cost_units,
                "failover_used":  r.failover_used,
                "failover_from":  r.failover_from,
                "failover_reason": r.failover_reason,
                "dyn_threshold":  r.dyn_threshold,
                "dyn_cost_weight": r.dyn_cost_weight,
                "response":       r.response,
            }
            for r in self.sub_results
        ]

    def to_ui_payload(self) -> dict:
        return {
            "prompt":    self.prompt,
            "total_ms":  self.total_ms,
            "merged":    self.merged,
            "sub_tasks": self.routing_table(),
        }


class Orchestrator:
    def __init__(
        self,
        feedback_store: Optional[FeedbackStore] = None,
        event_queue:    Optional[asyncio.Queue]  = None,
    ) -> None:
        self._store   = feedback_store or FeedbackStore()
        self._eq      = event_queue
        if event_queue:
            telemetry_engine.set_event_queue(event_queue)

    def set_event_queue(self, q: asyncio.Queue) -> None:
        self._eq = q
        telemetry_engine.set_event_queue(q)

    async def handle(self, prompt: str) -> OrchestratorResult:
        t_start  = time.perf_counter()
        profiles = split_and_classify(prompt)
        routing  = dispatch(profiles)

        # Broadcast routing plan before any LLM call
        await self._emit({
            "type":     "routing_plan",
            "prompt":   prompt,
            "sub_tasks": [
                {
                    "task_type":  p.task_type.value,
                    "complexity": p.complexity,
                    "tier":       p.tier.name,
                    "model":      m.label,
                    "provider":   m.provider,
                    "failover":   fe.fallback_label if fe else None,
                }
                for _, (p, m, fe) in routing.items()
            ],
        })

        async def _exec(
            key: str,
            profile: TaskProfile,
            model: ModelSpec,
            failover_event: Optional[FailoverEvent],
        ) -> SubTaskResult:

            # Emit "in_progress" event
            await self._emit({
                "type":       "subtask_start",
                "key":        key,
                "task_type":  profile.task_type.value,
                "model":      model.label,
                "provider":   model.provider,
                "failover":   failover_event.fallback_label if failover_event else None,
            })

            response, latency, tokens = await call_model(
                model_label=model.label,
                prompt=profile.raw_text,
                max_tokens=min(2048, profile.token_estimate * 3),
            )

            lat_ms  = int(latency * 1000)
            success = not response.startswith("[LiteLLM ERROR")

            # Feed telemetry engine (triggers threshold recomputation)
            quality = FeedbackStore._heuristic_quality(lat_ms, tokens)
            tel_rec = TelemetryRecord(
                model_label=model.label,
                task_type=profile.task_type.value,
                latency_ms=lat_ms,
                out_tokens=tokens,
                cost_units=round(model.cost_per_1k * tokens / 1000, 6),
                quality_eval=quality,
                success=success,
            )
            dyn_params = telemetry_engine.record(tel_rec)

            u          = _utility(model, profile)
            q          = model.quality_score(profile.complexity)
            thr        = STATIC_TASK_THRESHOLDS.get(profile.task_type.value, BASE_QUALITY_THRESHOLD)
            ollama_name = TIER_TO_MODEL.get(model.label, model.label)
            cost        = round(model.cost_per_1k * tokens / 1000, 6)

            self._store.record(
                task_type=profile.task_type.value,
                complexity=profile.complexity,
                tier=model.tier,
                model_used=model.label,
                latency_ms=lat_ms,
                output_tokens=tokens,
                cost_per_1k=model.cost_per_1k,
                prompt_len=len(profile.raw_text),
                failover_used=failover_event is not None,
                failover_from=failover_event.primary_label if failover_event else "",
                failover_reason=failover_event.reason if failover_event else "",
            )

            result = SubTaskResult(
                profile=profile,
                model=model,
                response=response,
                latency_ms=lat_ms,
                out_tokens=tokens,
                utility_score=round(u, 4),
                quality_score=round(q, 4),
                threshold=thr,
                ollama_model=ollama_name,
                cost_units=cost,
                failover_used=failover_event is not None,
                failover_from=failover_event.primary_label if failover_event else "",
                failover_reason=failover_event.reason if failover_event else "",
                dyn_threshold=dyn_params.quality_threshold,
                dyn_cost_weight=dyn_params.cost_weight,
            )

            # Emit "complete" event with full metadata
            await self._emit({
                "type":         "subtask_complete",
                "key":          key,
                "task_type":    profile.task_type.value,
                "model":        model.label,
                "provider":     model.provider,
                "tier":         model.tier,
                "latency_ms":   lat_ms,
                "out_tokens":   tokens,
                "cost_units":   cost,
                "quality":      round(q, 4),
                "utility":      round(u, 4),
                "threshold":    thr,
                "dyn_threshold": dyn_params.quality_threshold,
                "dyn_cost_weight": dyn_params.cost_weight,
                "success":      success,
                "failover_used": failover_event is not None,
                "failover_from": failover_event.primary_label if failover_event else "",
                "failover_reason": failover_event.reason if failover_event else "",
                "health":       failover_manager.health_snapshot(),
            })

            return result

        sub_results = list(await asyncio.gather(
            *[_exec(k, p, m, fe) for k, (p, m, fe) in routing.items()]
        ))

        merged   = self._merge(sub_results)
        total_ms = int((time.perf_counter() - t_start) * 1000)

        await self._emit({
            "type":     "request_complete",
            "total_ms": total_ms,
            "merged":   merged,
        })

        return OrchestratorResult(
            prompt=prompt,
            sub_results=sub_results,
            merged=merged,
            total_ms=total_ms,
        )

    def handle_sync(self, prompt: str) -> OrchestratorResult:
        return asyncio.run(self.handle(prompt))

    async def _emit(self, payload: dict) -> None:
        if self._eq is not None:
            try:
                self._eq.put_nowait(payload)
            except asyncio.QueueFull:
                pass

    @staticmethod
    def _merge(results: list[SubTaskResult]) -> str:
        sections = []
        for r in results:
            label  = r.profile.task_type.value.replace("_", " ").title()
            fo_tag = f" ⚡ failover from {r.failover_from}" if r.failover_used else ""
            header = (
                f"## {label}  "
                f"_(via {r.model.label} [{r.model.provider}]{fo_tag}, {r.latency_ms} ms)_"
            )
            sections.append(f"{header}\n\n{r.response.strip()}")
        return "\n\n---\n\n".join(sections)

File: v5-go-implementation/internal/orchestrator/orchestrator.go
// Package orchestrator ties together splitting → routing → parallel execution
// → telemetry recording → feedback persistence → WebSocket broadcast.
//
// Every completed sub-task emits an event to the eventCh channel so the
// front-end receives real-time updates.
package orchestrator

import (
	"context"
	"fmt"
	"math"
	"strings"
	"sync"
	"time"
	"unicode"

	"github.com/routing-v5/internal/classifier"
	"github.com/routing-v5/internal/failover"
	"github.com/routing-v5/internal/feedback"
	"github.com/routing-v5/internal/llmclient"
	"github.com/routing-v5/internal/registry"
	"github.com/routing-v5/internal/router"
	"github.com/routing-v5/internal/telemetry"
)

// SubTaskResult holds the outcome of one routed sub-task.
type SubTaskResult struct {
	Profile        classifier.TaskProfile
	Model          *registry.ModelSpec
	Response       string
	LatencyMs      int
	OutTokens      int
	UtilityScore   float64
	QualityScore   float64
	Threshold      float64
	OllamaModel    string
	CostUnits      float64
	FailoverUsed   bool
	FailoverFrom   string
	FailoverReason string
	DynThreshold   float64
	DynCostWeight  float64
}

// OrchestratorResult holds the full result of one /api/chat request.
type OrchestratorResult struct {
	Prompt     string
	SubResults []SubTaskResult
	Merged     string
	TotalMs    int
}

func (r *OrchestratorResult) RoutingTable() []map[string]interface{} {
	out := make([]map[string]interface{}, len(r.SubResults))
	for i, s := range r.SubResults {
		out[i] = map[string]interface{}{
			"task_type":        string(s.Profile.TaskType),
			"complexity":       s.Profile.Complexity,
			"tier":             s.Profile.Tier.String(),
			"model":            s.Model.Label,
			"provider":         s.Model.Provider,
			"ollama_model":     s.OllamaModel,
			"utility":          r4(s.UtilityScore),
			"quality":          r4(s.QualityScore),
			"threshold":        s.Threshold,
			"latency_ms":       s.LatencyMs,
			"out_tokens":       s.OutTokens,
			"cost_units":       s.CostUnits,
			"failover_used":    s.FailoverUsed,
			"failover_from":    s.FailoverFrom,
			"failover_reason":  s.FailoverReason,
			"dyn_threshold":    s.DynThreshold,
			"dyn_cost_weight":  s.DynCostWeight,
			"response":         s.Response,
		}
	}
	return out
}

func (r *OrchestratorResult) ToUIPayload() map[string]interface{} {
	return map[string]interface{}{
		"prompt":    r.Prompt,
		"total_ms":  r.TotalMs,
		"merged":    r.Merged,
		"sub_tasks": r.RoutingTable(),
	}
}

// Orchestrator is the central coordinator.
type Orchestrator struct {
	store   *feedback.Store
	eng     *telemetry.Engine
	fm      *failover.Manager
	eventCh chan<- map[string]interface{}
}

// New creates an Orchestrator.
func New(
	store *feedback.Store,
	eng *telemetry.Engine,
	fm *failover.Manager,
) *Orchestrator {
	return &Orchestrator{store: store, eng: eng, fm: fm}
}

// SetEventChannel registers the channel for WebSocket broadcast events.
func (o *Orchestrator) SetEventChannel(ch chan<- map[string]interface{}) {
	o.eventCh = ch
	o.eng.SetEventChannel(ch)
}

// Handle processes a prompt end-to-end and returns the full result.
func (o *Orchestrator) Handle(ctx context.Context, prompt string) (*OrchestratorResult, error) {
	tStart := time.Now()
	profiles := classifier.SplitAndClassify(prompt)
	routing := router.Dispatch(o.eng, o.fm, profiles)

	// Broadcast routing plan
	planTasks := make([]map[string]interface{}, 0, len(routing))
	for _, dr := range routing {
		entry := map[string]interface{}{
			"task_type":  string(dr.Profile.TaskType),
			"complexity": dr.Profile.Complexity,
			"tier":       dr.Profile.Tier.String(),
			"model":      dr.Model.Label,
			"provider":   dr.Model.Provider,
		}
		if dr.FailoverEvent != nil {
			entry["failover"] = dr.FailoverEvent.FallbackLabel
		}
		planTasks = append(planTasks, entry)
	}
	o.emit(map[string]interface{}{
		"type":      "routing_plan",
		"prompt":    prompt,
		"sub_tasks": planTasks,
	})

	// Execute all sub-tasks in parallel
	type indexedResult struct {
		key string
		res SubTaskResult
		err error
	}
	ch := make(chan indexedResult, len(routing))
	var wg sync.WaitGroup

	for key, dr := range routing {
		wg.Add(1)
		go func(key string, dr router.DispatchResult) {
			defer wg.Done()
			res, err := o.execSubTask(ctx, key, dr)
			ch <- indexedResult{key: key, res: res, err: err}
		}(key, dr)
	}

	wg.Wait()
	close(ch)

	var subResults []SubTaskResult
	for ir := range ch {
		if ir.err == nil {
			subResults = append(subResults, ir.res)
		}
	}

	merged := merge(subResults)
	totalMs := int(time.Since(tStart).Milliseconds())

	o.emit(map[string]interface{}{
		"type":     "request_complete",
		"total_ms": totalMs,
		"merged":   merged,
	})

	return &OrchestratorResult{
		Prompt:     prompt,
		SubResults: subResults,
		Merged:     merged,
		TotalMs:    totalMs,
	}, nil
}

func (o *Orchestrator) execSubTask(
	ctx context.Context,
	key string,
	dr router.DispatchResult,
) (SubTaskResult, error) {
	profile := dr.Profile
	model := dr.Model
	fev := dr.FailoverEvent

	failoverLabel := ""
	if fev != nil {
		failoverLabel = fev.FallbackLabel
	}

	o.emit(map[string]interface{}{
		"type":      "subtask_start",
		"key":       key,
		"task_type": string(profile.TaskType),
		"model":     model.Label,
		"provider":  model.Provider,
		"failover":  failoverLabel,
	})

	maxTokens := int(math.Min(2048, float64(profile.TokenEstimate*3)))
	cr := llmclient.CallModel(ctx, o.fm, model.Label, profile.RawText, maxTokens)

	latMs := int(cr.LatencySec * 1000)
	success := !strings.HasPrefix(cr.Text, "[LiteLLM ERROR")

	quality := feedback.HeuristicQuality(latMs, cr.Tokens)
	telRec := telemetry.Record{
		ModelLabel:  model.Label,
		TaskType:    string(profile.TaskType),
		LatencyMs:   latMs,
		OutTokens:   cr.Tokens,
		CostUnits:   r6(model.CostPer1k * float64(cr.Tokens) / 1000),
		QualityEval: quality,
		Success:     success,
	}
	dynParams := o.eng.Record(telRec)

	u := router.Utility(o.eng, model, profile)
	q := model.QualityAt(profile.Complexity)
	thr := router.StaticTaskThresholds[string(profile.TaskType)]
	if thr == 0 {
		thr = router.BaseQualityThreshold
	}
	ollamaName := llmclient.TierModels[model.Label]
	if ollamaName == "" {
		ollamaName = model.Label
	}
	cost := r6(model.CostPer1k * float64(cr.Tokens) / 1000)

	failoverFrom := ""
	failoverReason := ""
	if fev != nil {
		failoverFrom = fev.PrimaryLabel
		failoverReason = fev.Reason
	}

	o.store.Record(
		string(profile.TaskType),
		profile.Complexity,
		model.Tier,
		model.Label,
		latMs,
		cr.Tokens,
		model.CostPer1k,
		len(profile.RawText),
		fev != nil,
		failoverFrom,
		failoverReason,
	)

	res := SubTaskResult{
		Profile:        profile,
		Model:          model,
		Response:       cr.Text,
		LatencyMs:      latMs,
		OutTokens:      cr.Tokens,
		UtilityScore:   r4(u),
		QualityScore:   r4(q),
		Threshold:      thr,
		OllamaModel:    ollamaName,
		CostUnits:      cost,
		FailoverUsed:   fev != nil,
		FailoverFrom:   failoverFrom,
		FailoverReason: failoverReason,
		DynThreshold:   dynParams.QualityThreshold,
		DynCostWeight:  dynParams.CostWeight,
	}

	health := o.fm.HealthSnapshot()
	o.emit(map[string]interface{}{
		"type":            "subtask_complete",
		"key":             key,
		"task_type":       string(profile.TaskType),
		"model":           model.Label,
		"provider":        model.Provider,
		"tier":            model.Tier,
		"latency_ms":      latMs,
		"out_tokens":      cr.Tokens,
		"cost_units":      cost,
		"quality":         r4(q),
		"utility":         r4(u),
		"threshold":       thr,
		"dyn_threshold":   dynParams.QualityThreshold,
		"dyn_cost_weight": dynParams.CostWeight,
		"success":         success,
		"failover_used":   fev != nil,
		"failover_from":   failoverFrom,
		"failover_reason": failoverReason,
		"health":          health,
	})

	return res, nil
}

func (o *Orchestrator) emit(payload map[string]interface{}) {
	if o.eventCh != nil {
		select {
		case o.eventCh <- payload:
		default:
		}
	}
}

func merge(results []SubTaskResult) string {
	sections := make([]string, 0, len(results))
	for _, r := range results {
		label := strings.ReplaceAll(string(r.Profile.TaskType), "_", " ")
		// title-case without deprecated strings.Title
		words := strings.Fields(label)
		for i, w := range words {
			if len(w) > 0 {
				runes := []rune(w)
				runes[0] = unicode.ToTitle(runes[0])
				words[i] = string(runes)
			}
		}
		label = strings.Join(words, " ")
		foTag := ""
		if r.FailoverUsed {
			foTag = fmt.Sprintf(" ⚡ failover from %s", r.FailoverFrom)
		}
		header := fmt.Sprintf("## %s  _(via %s [%s]%s, %d ms)_",
			label, r.Model.Label, r.Model.Provider, foTag, r.LatencyMs)
		sections = append(sections, header+"\n\n"+strings.TrimSpace(r.Response))
	}
	return strings.Join(sections, "\n\n---\n\n")
}

func r4(f float64) float64  { return math.Round(f*10000) / 10000 }
func r6(f float64) float64  { return math.Round(f*1e6) / 1e6 }
